packages/application-shell-connectors/src/components/application-context/normalizers.ts

```diff
@@ -0,0 +1,218 @@
+  [key: string]: TStoreDataFence[];
+};
+
+const normalizeAppliedDataFencesForStoresByResourceType = (
+  dataFences: Maybe<TStoreDataFence>[]
+) => {
+  const groupedByResourceType = dataFences.reduce<
+    NormalizedStoresGroupByResourceType
+  >((previousGroupsOfSameType, appliedDataFence) => {
+    if (!appliedDataFence) return previousGroupsOfSameType;
+    const previousGroup = previousGroupsOfSameType[appliedDataFence.group];
+    return {
+      ...previousGroupsOfSameType,
+      [appliedDataFence.group]: [...(previousGroup || []), appliedDataFence],
+    };
+  }, {});
+  return Object.entries(groupedByResourceType).reduce<
+    NormalizedGroupedByResourceType
+  >((previousGroupedByResourceType, [resourceType, dataFences]) => {
+    const groupByDataFenceName = dataFences.reduce<
+      NormalizedGroupedByPermission
+    >((nextDataFenceValues, dataFence) => {
+      const dataFenceByName = nextDataFenceValues[dataFence.name] || {
+        values: [],
+      };
+      return {
+        ...nextDataFenceValues,
+        [dataFence.name]: {
+          ...dataFenceByName,
+          values: [...dataFenceByName.values, dataFence.value],
+        },
+      };
+    }, {});
+    return {
+      ...previousGroupedByResourceType,
+      [resourceType]: groupByDataFenceName,
+    };
+  }, {});
+};
+
+// input:
+// [
+//   {
+//     type: 'store',
+//     name: 'canManageOrders',
+//     value: 'usa',
+//     group: 'orders',
+//   },
+//   {
+//     type: 'store',
+//     name: 'canManageOrders',
+//     value: 'germany',
+//     group: 'orders',
+//   },
+//   {
+//     type: 'store',
+//     name: 'canViewOrders',
+//     value: 'canada',
+//     group: 'orders',
+//   },
+// ]
+// output:
+// {
+//   store: {
+//     orders: {
+//       canManageOrders: { values: ['usa', 'germany'] },
+//       canViewOrders: { values: ['canada'] },
+//     }
+//   }
+// }
+
+type NormalizedGroupByType = {
+  [key: string]: Maybe<TStoreDataFence>[];
+};
+
+export const normalizeAllAppliedDataFences = (
+  project: TFetchProjectQuery['project']
+) => {
+  if (!project) return null;
+  const allAppliedDataFences = project.allAppliedDataFences;
+  if (allAppliedDataFences.length === 0) {
+    return null;
+  }
+  const groupedByType = allAppliedDataFences.reduce<NormalizedGroupByType>(
+    (previousGroupsOfSameType, appliedDataFence) => {
+      if (!appliedDataFence) return previousGroupsOfSameType;
+      const previousGroup = previousGroupsOfSameType[appliedDataFence.type];
+      return {
+        ...previousGroupsOfSameType,
+        [appliedDataFence.type]: [...(previousGroup || []), appliedDataFence],
+      };
+    },
+    {}
+  );
+  const normalizedDataFences = {
+    ...('store' in groupedByType
+      ? {
+          store: normalizeAppliedDataFencesForStoresByResourceType(
+            groupedByType.store
+          ),
+        }
+      : {}),
+  };
+  return Object.keys(normalizedDataFences).length > 0
+    ? normalizedDataFences
+    : null;
+};
```

Comment:
This is nice. I assume the contents/behaviour here didn't change. It's hard/impossible to trace those changes without very small commits while renaming and moving happens.

Reply:
Yeah, it basically didn't change, I only added the type annotations